quickwit/quickwit-search/src/list_fields.rs

```diff
@@ -0,0 +1,623 @@
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::{HashMap, HashSet};
+use std::io;
+use std::path::Path;
+use std::sync::Arc;
+
+use futures::future::try_join_all;
+use itertools::Itertools;
+use quickwit_common::shared_consts::SPLIT_FIELDS_FILE_NAME;
+use quickwit_common::uri::Uri;
+use quickwit_indexing::models::read_split_fields;
+use quickwit_metastore::{ListIndexesMetadataResponseExt, SplitMetadata};
+use quickwit_proto::metastore::{
+    ListIndexesMetadataRequest, MetastoreService, MetastoreServiceClient,
+};
+use quickwit_proto::search::{
+    LeafListFieldsRequest, ListFieldsEntryResponse, ListFieldsRequest, ListFieldsResponse,
+    SplitIdAndFooterOffsets,
+};
+use quickwit_proto::types::IndexUid;
+use quickwit_storage::Storage;
+use tantivy::FieldMetadata;
+
+use crate::leaf::open_split_bundle;
+use crate::service::SearcherContext;
+use crate::{list_relevant_splits, ClusterClient, SearchError, SearchJob};
+
+/// Get the list of splits for the request which we need to scan.
+pub async fn get_fields_from_split<'a>(
+    searcher_context: &SearcherContext,
+    index_id: String,
+    split_and_footer_offsets: &'a SplitIdAndFooterOffsets,
+    index_storage: Arc<dyn Storage>,
+) -> anyhow::Result<impl Iterator<Item = io::Result<ListFieldsEntryResponse>>> {
+    // TODO: Add fancy caching
+    let (_, split_bundle) =
+        open_split_bundle(searcher_context, index_storage, split_and_footer_offsets).await?;
+
+    let serialized_split_fields = split_bundle
+        .get_all(Path::new(SPLIT_FIELDS_FILE_NAME))
+        .await?;
+    let serialized_split_fields_len = serialized_split_fields.len();
+    let iter = read_split_fields(serialized_split_fields).map_err(|err| {
+        anyhow::anyhow!(
+            "could not read split fields (serialized len: {}): {:?}",
+            serialized_split_fields_len,
+            err
+        )
+    })?;
+
+    Ok(iter.map(move |metadata| {
+        metadata.map(|metadata| ListFieldsEntryResponse {
+            field_name: metadata.field_name,
+            field_type: metadata.typ.to_code() as i32,
+            index_ids: vec![index_id.to_string()],
+            searchable: metadata.indexed,
+            aggregatable: metadata.fast,
+            non_searchable_index_ids: Vec::new(),
```

Comment:
I don't understand this contract. 
So you return io::Result and lose the info of the missing split ids? 
What are non_searchable_index_ids ?

Reply:
When do we return an ioError as an item.

Reply:
Actually I'm not sure what to do with the error, currently we abort and return. This is happening in the merge function. I guess we want to return the failed split_ids for a retry.
```rust
let entry = entry.map_err(|err| crate::error::SearchError::Internal(format!("{:?}", err)))?; // TODO: No early return on error
```

`non_searchable_index_ids` are exemption list of indexes if `searchable: true`. I took this from the elastic search API.